lib/tdf3/src/tdf.ts

```diff
@@ -1107,6 +1089,47 @@ export class TDF extends EventEmitter {
     }
   }
 
+  async sliceAndDecrypt({
+    buffer,
+    reconstructedKeyBinary,
+    slice,
+  }: {
+    buffer: Uint8Array;
+    reconstructedKeyBinary: Binary;
+    slice: Chunk[];
+  }) {
+    const batchSize = 10;
+    let promises = [];
+
+    for (const index in slice) {
+      const { encryptedOffset, encryptedSegmentSize } = slice[index];
+
+      const offset =
+        slice[0].encryptedOffset === 0
+          ? encryptedOffset
+          : encryptedOffset % slice[0].encryptedOffset;
+      const encryptedChunk = new Uint8Array(
+        buffer.slice(offset, offset + (encryptedSegmentSize as number))
+      );
+
+      promises.push(
+        this.decryptChunk(encryptedChunk, reconstructedKeyBinary, slice[index]['hash']).then(
+          (decryptedChunk) => {
+            if (!decryptedChunk) return;
+            slice[index].decryptedChunk = decryptedChunk;
+            if (slice[index]._resolve) {
+              (slice[index]._resolve as (value: unknown) => void)(null);
+            }
+          }
+        )
+      );
+      if (promises.length === batchSize) {
```

Comment:
There is a bug here - it won't wait for all batches if slice.length isn't a multiple of batchSize.

Suggestion: Use `p-limit` instead: https://www.npmjs.com/package/p-limit

This has the benefit of not batching, so may even be faster, and should simplify this code overall